handlers/stream.go

```diff
@@ -0,0 +1,180 @@
+package handlers
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/companieshouse/chs-go-avro-schemas/data"
+	"github.com/companieshouse/chs-streaming-api-frontend/consumer"
+	"github.com/companieshouse/chs-streaming-api-frontend/offset"
+
+	"github.com/companieshouse/chs.go/log"
+	"github.com/gorilla/pat"
+)
+
+// Stream should be implemented by the different stream types
+type Stream interface {
+	// Transform modifies data if required within this microservice
+	Transform(req *http.Request, msg *[]byte) error
+	// Filter modifies the data to a given filtered view set in the request
+	Filter(req *http.Request, msg *[]byte) error
+}
+
+//Streaming contains necessary config for streaming
+type Streaming struct {
+	BrokerAddr        []string
+	RequestTimeout    time.Duration
+	HeartbeatInterval time.Duration
+	//CacheBroker:       s,
+	ConsumerFactory consumer.FactoryInterface
+	Offset          offset.Interface
+}
+
+// AddStream sets up the routing for the particular stream type
+func (st Streaming) AddStream(router *pat.Router, route string, topic string, stream Stream) {
+	router.Path(route).Methods("GET").HandlerFunc(st.process(topic, stream))
+}
+
+func (st Streaming) process(topic string, stream Stream) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+
+		log.InfoC(req.Header.Get("ERIC_Identity"), "consuming from topic", log.Data{"topic": topic})
+
+		st.ProcessHTTP(topic, w, req, stream)
+	}
+}
+
+/*
+  Have mini functions for handling timer events so that we can stub these in tests to verify that they are called
+*/
+func handleTimeOut(contextID string) {
+	log.DebugC(contextID, "connection expired, disconnecting client ")
+}
+
+func handleClientDisconnect(contextID string) {
+	log.DebugC(contextID, "connection closed by client")
+}
+
+func logHeartbeatTimeout(contextID string) {
+	log.TraceC(contextID, "application heartbeat")
+}
+
+var callLogHandleTimeOut = handleTimeOut
+var callLogHandleClientDisconnect = handleClientDisconnect
+var callProcessMessage = ProcessMessage
+var callProcessMessageFailed = processMessageFailed
+var callLogHeartbeatTimeout = logHeartbeatTimeout
+var callCheckIfDataNotPresent = checkIfDataNotPresent
+var callResponseWriterWrite = responseWriterWrite
+
+// ProcessHTTP Test at this level for unit testing.
+func (st Streaming) ProcessHTTP(topic string, w http.ResponseWriter, req *http.Request, stream Stream) {
```

Comment:
Given that the frontend will now be decoupled from Kafka, this handler should do two things:

1. Fetch a range of offsets cached by chs-streaming-api-cache if a timepoint has been specified.
2. Stream offsets published to it by chs-streaming-api-cache to the connected user.

Reply:
noted and added as comment for the next PR